File: new-backend/app/Models/Product.php
<?php

namespace App\Models;

use App\Services\SlugService;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Product extends Model
{
    use HasFactory;

    protected $table = 'products';

    protected $fillable = [
        'name',
        'slug',
        'description',
        'price',
        'discount_price',
        'stock',
        'category_id',
        'image_url',
        'status',
    ];

    protected function casts(): array
    {
        return [
            'price' => 'decimal:2',
            'discount_price' => 'decimal:2',
            'stock' => 'integer',
            'category_id' => 'integer',
        ];
    }

    protected static function booted(): void
    {
        static::saving(function (Product $product) {
            if (empty($product->slug) || $product->isDirty('name')) {
                $product->slug = app(SlugService::class)->uniqueFor(
                    'products',
                    $product->name ?? '',
                    $product->exists ? $product->id : null
                );
            }
        });
    }

    public function category(): BelongsTo
    {
        return $this->belongsTo(Category::class);
    }

    public function images(): HasMany
    {
        return $this->hasMany(ProductImage::class)->orderBy('sort_order')->orderBy('id');
    }

    public function scopeActive(Builder $query): Builder
    {
        return $query->where('status', 'active');
    }

    public function effectivePrice(): float
    {
        $discount = $this->discount_price;

        return $discount !== null && (float) $discount > 0
            ? (float) $discount
            : (float) $this->price;
    }
}

File: mobile/lib/models/product.dart
class ProductImage {
  ProductImage({
    required this.id,
    required this.url,
    required this.sortOrder,
  });

  final int id;
  final String url;
  final int sortOrder;

  factory ProductImage.fromJson(Map<String, dynamic> j) {
    return ProductImage(
      id: (j['id'] as num).toInt(),
      url: j['url'] as String? ?? '',
      sortOrder: (j['sort_order'] as num?)?.toInt() ?? 0,
    );
  }
}

class Product {
  Product({
    required this.id,
    required this.name,
    required this.slug,
    this.description,
    required this.price,
    this.discountPrice,
    required this.stock,
    required this.categoryId,
    this.categoryName,
    this.categorySlug,
    this.imageUrl,
    required this.status,
    this.images = const [],
    this.createdAt,
    this.updatedAt,
  });

  final int id;
  final String name;
  final String slug;
  final String? description;
  final double price;
  final double? discountPrice;
  final int stock;
  final int categoryId;
  final String? categoryName;
  final String? categorySlug;
  final String? imageUrl;
  final String status;
  final List<ProductImage> images;
  final String? createdAt;
  final String? updatedAt;

  factory Product.fromJson(Map<String, dynamic> j) {
    final imgs = j['images'];
    List<ProductImage> list = [];
    if (imgs is List) {
      list = imgs
          .map((e) => ProductImage.fromJson(Map<String, dynamic>.from(e as Map)))
          .toList();
    }
    return Product(
      id: (j['id'] as num).toInt(),
      name: j['name'] as String? ?? '',
      slug: j['slug'] as String? ?? '',
      description: j['description'] as String?,
      price: (j['price'] as num?)?.toDouble() ?? 0,
      discountPrice: (j['discount_price'] as num?)?.toDouble(),
      stock: (j['stock'] as num?)?.toInt() ?? 0,
      categoryId: (j['category_id'] as num?)?.toInt() ?? 0,
      categoryName: j['category_name'] as String?,
      categorySlug: j['category_slug'] as String?,
      imageUrl: j['image_url'] as String?,
      status: j['status'] as String? ?? 'active',
      images: list,
      createdAt: j['created_at'] as String?,
      updatedAt: j['updated_at'] as String?,
    );
  }
}
